subsys/bluetooth/controller/ll_sw/ull_adv_sync.c

```diff
@@ -507,6 +521,151 @@ void ull_adv_sync_offset_get(struct ll_adv_set *adv)
 	LL_ASSERT(!ret);
 }
 
+uint8_t ull_adv_sync_acad_enable(struct lll_adv_sync *lll_sync,
+				 uint8_t acad_len, void **acad)
+{
+	struct pdu_adv_com_ext_adv *ter_com_hdr, *ter_com_hdr_prev;
+	struct pdu_adv_ext_hdr *ter_hdr, ter_hdr_prev;
+	struct pdu_adv *ter_pdu, *ter_pdu_prev;
+	uint8_t *ter_dptr, *ter_dptr_prev;
+	uint16_t ter_len, ter_len_prev;
+	uint8_t acad_len_prev;
+	uint8_t *ad_data;
+	uint8_t ter_idx;
+	uint8_t ad_len;
+
+	/* Get reference to previous tertiary PDU data */
+	ter_pdu_prev = lll_adv_sync_data_peek(lll_sync);
+	ter_com_hdr_prev = (void *)&ter_pdu_prev->adv_ext_ind;
+	ter_hdr = (void *)ter_com_hdr_prev->ext_hdr_adv_data;
+	ter_hdr_prev = *ter_hdr;
+	ter_dptr_prev = ter_hdr->data;
+
+	/* Get reference to new tertiary PDU data buffer */
+	ter_pdu = lll_adv_sync_data_alloc(lll_sync, &ter_idx);
+	ter_pdu->type = ter_pdu_prev->type;
+	ter_pdu->rfu = 0U;
+	ter_pdu->chan_sel = 0U;
+	ter_com_hdr = (void *)&ter_pdu->adv_ext_ind;
+	ter_com_hdr->adv_mode = ter_com_hdr_prev->adv_mode;
+	ter_hdr = (void *)ter_com_hdr->ext_hdr_adv_data;
+	ter_dptr = ter_hdr->data;
+	*(uint8_t *)ter_hdr = 0U;
+
+	/* No AdvA */
+	/* No TargetA */
+
+	/* TODO: CTEInfo */
+
+	/* No ADI */
+
+	/* AuxPtr */
+	if (ter_hdr_prev.aux_ptr) {
+		ter_dptr_prev += sizeof(struct pdu_adv_aux_ptr);
+
+		ter_hdr->aux_ptr = 1;
+		ter_dptr += sizeof(struct pdu_adv_aux_ptr);
+	}
+
+	/* No SyncInfo */
+
+	/* Tx Power flag */
+	if (ter_hdr_prev.tx_pwr) {
+		ter_dptr_prev++;
+
+		ter_hdr->tx_pwr = 1;
+		ter_dptr++;
+	}
+
+	/* Calc previous ACAD len and update PDU len */
+	ter_len_prev = ter_dptr_prev - (uint8_t *)ter_com_hdr_prev;
+	if (ter_len_prev < (ter_com_hdr_prev->ext_hdr_len +
+			    offsetof(struct pdu_adv_com_ext_adv,
+				     ext_hdr_adv_data))) {
+		acad_len_prev = ter_com_hdr_prev->ext_hdr_len +
+				offsetof(struct pdu_adv_com_ext_adv,
+					 ext_hdr_adv_data) - ter_len_prev;
+		ter_len_prev += acad_len_prev;
+		ter_dptr_prev += acad_len_prev;
+	} else {
+		acad_len_prev = 0;
+		ter_len_prev = offsetof(struct pdu_adv_com_ext_adv,
+					ext_hdr_adv_data);
+		ter_dptr_prev = (uint8_t *)ter_com_hdr_prev + ter_len_prev;
+	}
+
+	/* Did we parse beyond PDU length? */
+	if (ter_len_prev > ter_pdu_prev->len) {
+		/* we should not encounter invalid length */
+		return BT_HCI_ERR_UNSPECIFIED;
+	}
```

Comment:
Do we need to free `ter_pdu` here?

Reply:
Would an ASSERT be better to use here then? It doesn't make much sense to have a runtime check if we are not going to handle it, because it's caused by an implementation error.